src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts

```diff
@@ -177,9 +177,6 @@ export class TableComponent implements AfterContentChecked, OnInit, OnChanges, O
   }
 
   ngOnInit() {
-    // ngx-datatable triggers calculations each time mouse enters a row,
-    // this will prevent that.
-    window.addEventListener('mouseenter', (event) => event.stopPropagation(), true);
```

Comment:
Can you please try if this will solve the problem for your:
```
--- src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts	(revision 45b017f8cd5ddbc4a65c4f8dcedd099e738bc3a7)
+++ src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts	(date 1568195189090)
@@ -3,6 +3,7 @@
   ChangeDetectionStrategy,
   ChangeDetectorRef,
   Component,
+  ElementRef,
   EventEmitter,
   Input,
   NgZone,
@@ -164,7 +165,11 @@
   // table columns after the browser window has been resized.
   private currentWidth: number;
 
-  constructor(private ngZone: NgZone, private cdRef: ChangeDetectorRef) {}
+  constructor(
+    private elementRef: ElementRef,
+    private ngZone: NgZone,
+    private cdRef: ChangeDetectorRef
+  ) {}
 
   static prepareSearch(search: string) {
     search = search.toLowerCase().replace(/,/g, '');
@@ -179,7 +184,11 @@
   ngOnInit() {
     // ngx-datatable triggers calculations each time mouse enters a row,
     // this will prevent that.
-    window.addEventListener('mouseenter', (event) => event.stopPropagation(), true);
+    this.elementRef.nativeElement.addEventListener(
+      'mouseenter',
+      this.onMouseEnter.bind(this),
+      true
+    );
     this._addTemplates();
     if (!this.sorts) {
       // Check whether the specified identifier exists.
@@ -307,6 +316,10 @@
     this.tableColumns = this.columns.filter((c) => !c.isHidden);
   }
 
+  onMouseEnter(event) {
+    event.stopPropagation();
+  }
+
   ngOnDestroy() {
     if (this.reloadSubscriber) {
       this.reloadSubscriber.unsubscribe();
@@ -314,6 +327,7 @@
     if (this.saveSubscriber) {
       this.saveSubscriber.unsubscribe();
     }
+    this.elementRef.nativeElement.removeEventListener('mouseenter', this.onMouseEnter);
   }
 
   ngAfterContentChecked() {
```
IMO the problem of the current implementation is that the listener is added to the ``windows`` object instead of the datatable itself.

Reply:
Your suggested changes seem to work fine on Firefox. I haven't had any performance issues when testing this patch with Firefox. Thank you!

But it seems that removing the workaround (that's what this PR does) to fix the Firefox performance issue, causes another issue with another workaround. And it does not matter if I just removed the workaround like done in this PR or applied your changes to adapt it to perform better.

```
Uncaught Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'innerWidth: 1471'. Current value: 'innerWidth: 1459'
```

I initially thought that this error only occurs on Firefox, but was, in meanwhile, able to reproduce it on Chrome as well.

It seems to be related to the workaround provided for the `Empty datatable rendered before data has been fetched` [issue](https://tracker.ceph.com/issues/41615).

When I remove the content of `ngAfterContentChecked()` in `table.component.ts`, there's no such error in the console anymore. Unfortunately, the previous table rendering issue reappears.

Reply:
After doing some more tests and creating a new PR the following patch in combination with #30335 should work. Please give it a try.

```
Index: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts	(date 1568193542000)
+++ src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts	(date 1568206668571)
@@ -177,6 +177,13 @@
   }
 
   ngOnInit() {
+    // ngx-datatable triggers calculations each time mouse enters a row,
+    // this will prevent that.
+    this.table.element.addEventListener(
+      'mouseenter',
+      this.onMouseEnter.bind(this),
+      true
+    );
     this._addTemplates();
     if (!this.sorts) {
       // Check whether the specified identifier exists.
@@ -304,6 +311,10 @@
     this.tableColumns = this.columns.filter((c) => !c.isHidden);
   }
 
+  onMouseEnter(event) {
+    event.stopPropagation();
+  }
+
   ngOnDestroy() {
     if (this.reloadSubscriber) {
       this.reloadSubscriber.unsubscribe();
@@ -311,6 +322,7 @@
     if (this.saveSubscriber) {
       this.saveSubscriber.unsubscribe();
     }
+    this.table.element.removeEventListener('mouseenter', this.onMouseEnter);
   }
 
   ngAfterContentChecked() {
```

Reply:
Thanks, I adapted the last example and implemented it. I don't think it's necessary to remove the event listener of a DOM node that is just initialized once and removed with the event listener. So there's no need to remove it manually. That shortens the solution a bit.